APCAppCore/APCAppCore/Library/Scoring/APCScoring.m

```diff
@@ -575,113 +735,148 @@ - (NSArray *)retrieveResultSummaryFromResults:(NSSet *)results latestOnly:(BOOL)
         
-        for (NSDictionary *item in groupItems) {
-            NSNumber *value = [item valueForKey:kDatasetValueKey];
+        for (NSDictionary *dataPoint in pointsBetweenStartAndEndDates) {
+            NSNumber *value = [dataPoint valueForKey:kDatasetValueKey];
             
             if ([value integerValue] != NSNotFound) {
                 itemSum += [value doubleValue];
             }
         }
         
-        if (groupItems.count != 0) {
-            dayAverage = itemSum / groupItems.count;
+        if (pointsBetweenStartAndEndDates.count != 0) {
+            dayAverage = itemSum / pointsBetweenStartAndEndDates.count;
         }
         
         if (dayAverage == 0) {
             dayAverage = NSNotFound;
         }
         
-        // Set the min/max for the data
         APCRangePoint *rangePoint = [APCRangePoint new];
         
         if (dayAverage != NSNotFound) {
-            NSNumber *dataMinValue = [groupItems valueForKeyPath:@"@min.datasetValueKey"];
-            NSNumber *dataMaxValue = [groupItems valueForKeyPath:@"@max.datasetValueKey"];
+            NSNumber *minValue = [pointsBetweenStartAndEndDates valueForKeyPath:@"@min.datasetValueKey"];
+            NSNumber *maxValue = [pointsBetweenStartAndEndDates valueForKeyPath:@"@max.datasetValueKey"];
             
-            rangePoint.minimumValue = [dataMinValue floatValue];
-            rangePoint.maximumValue = [dataMaxValue floatValue];
+            rangePoint.minimumValue = [minValue floatValue];
+            rangePoint.maximumValue = [maxValue floatValue];
         }
         
-        [entry setObject:@(dayAverage) forKey:kDatasetValueKey];
-        [entry setObject:rangePoint forKey:kDatasetRangeValueKey];
+        NSMutableDictionary *entry = [[self generateDataPointForDate:key withValue:@(dayAverage) noDataValue:YES] mutableCopy];
+        entry[kDatasetRangeValueKey] = rangePoint;
         
-        [groupedDataset addObject:entry];
+        if (rawData) {
+            entry[kDatasetRawDataKey] = rawData;
+        }
+        
+        summarizedDataset[key] = @[entry];
     }
     
-    // resort the grouped dataset by date
-    NSSortDescriptor *sortByDate = [[NSSortDescriptor alloc] initWithKey:kDatasetDateKey
-                                                               ascending:YES];
-    [groupedDataset sortUsingDescriptors:@[sortByDate]];
-    
-    [self.dataPoints removeAllObjects];
-    [self.dataPoints addObjectsFromArray:groupedDataset];
+    return summarizedDataset;
 }
 
-- (void)groupDatasetbyPeriod:(APHTimelineGroups)period
+- (NSDictionary *)groupByKeyPath:(NSString *)key dataset:(NSArray *)dataset
 {
-    NSDateComponents *groupDateComponents = [[NSDateComponents alloc] init];
+    NSMutableDictionary *groups = [NSMutableDictionary new];
     
-    for (NSDate *groupStartDate in self.timeline) {
-        NSDate *groupEndDate   = nil;
+    for (id object in dataset) {
+        id value = [object valueForKeyPath:key];
```

Comment:
I'm not quite sure I follow why we're using integer values for specific keypaths for this, when we could much more easily use the timestamp on the object (which I hope exists) directly. Was this an attempt at a performance fix?